src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs

```diff
@@ -268,7 +268,13 @@ private void GenerateCheckBox(ModelExplorer modelExplorer, TagHelperOutput outpu
                 if (hiddenForCheckboxTag != null)
                 {
                     hiddenForCheckboxTag.TagRenderMode = renderingMode;
-                    output.Content.Append(hiddenForCheckboxTag);
+
+                    var hiddenForCheckboxList = ViewContext.FormContext.FormData.ContainsKey("hiddenForCheckBoxList") ?
+                        ViewContext.FormContext.FormData["hiddenForCheckBoxList"] as List<TagBuilder> :
+                        new List<TagBuilder>();
+
+                    hiddenForCheckboxList.Add(hiddenForCheckboxTag);
+                    ViewContext.FormContext.FormData["hiddenForCheckBoxList"] = hiddenForCheckboxList;
```

Comment:
This code should not be duplicated between here and `HtmlHelper`. Instead move it into `DefaultHtmlGenerator.GenerateHiddenForCheckbox()`, change that method to return `void`, and rename it to something like `SaveHiddenForCheckbox()`. This class and `HtmlHelper` will then get simpler rather than more complicated.